Add contents for `08 Basic App Porting` (#30)
unikraft/summer-of-code-2021: content/en/docs/sessions/08-basic-app-porting/index.md

but again, you must be able to access the source files and the build system
before you can begin.

For the sake of simplicity, this tutorial will only be targeting applications
which are C/C++-based.  Unikraft supports other compile-time languages, such as
Golang, Rust and WASM.  However, the scope of this tutorial only follows an
example with a C/++-based program.  Many of the principles in this tutorial,
however, can be applied in the same way for said languages, with a bit of
context-specific work.  Namely, this may include additional build rules for
target files, using specific compilers and linkers, etc.

It is worth noting that we are only targeting compile-time applications.
Runtime languages of applications, such as Python or Lua, require an interpreter
which must be brought to Unikraft first.  There are already lots of higher-level
languages supported by Unikraft.  Bringing a new high-level language to Unikraft
would also be in the scope of this tutorial, as the steps here would cover the
ones needed to bring the interpreter, which is a program after all, as a
Unikraft unikernel application.  In the case of higher-level languages which are

> interpreted, you do not need to follow this tutorial.  Instead, simply mount the

Author: I see what you mean.  My point is:

  * If you are wishing to run a HLL on Unikraft which is already supported, e.g. the reader wishes to "port a simple application written in python" (for example) then this is the wrong tutorial for the reader;
  * If you are wishing to port a new HLL interpreter to Unikraft, for example PHP, then this tutorial is for you.
  
  I'll update the text accordingly.

application code with the relevant Unikernel binary.  For example, mounting a
directory with python code to the python Unikraft unikernel.  Please review
[Session 03](/docs/sessions/04-complex-applications/) for more information on
this topic.

## Starting with a Linux userspace build

For the remainder of this tutorial, we will be targeting the network utility
program [`iperf3`](https://github.com/esnet/iperf) as our application example we
wish to bring to Unikraft.  `iperf3` is a benchmarking tool, and is used to
determine the bandwidth between a client and server.  It makes for an excellent
application to be run as a Unikernel because:

 * It can run as a "server-type" application, receiving and processing requests
   for clients;
 * It is a standalone tool which does one thing;
 * It's [GNU Make](https://www.gnu.org/software/make/) and C-based; and,
 * It's quite useful :)

> 

Author: Oh no! I am a big fan of `;` :D

   this:

   ```bash
   tree -L 1
   ```
   ```
   .
   ├── apps
   ├── archs
   ├── libs
   ├── plats
   └── unikraft

   5 directories, 0 files
   ```

2. Let's now create a library for `iperf3`.  We can use `kraft` to initialize
   some boilerplate for us too.  To do this, we must first retrieve some
   information about the program itself.  First we need to identify the latest
   version number of `iperf3`.  Github tells us (as of of time of writing this
   tutorial) that this is `3.10.1`.

   Unikraft relies on the ability to download the source code of the "origin"
   code which is about to be compiled.  Usually these are tarballs or zips.
   Ideally, we want to have a version number in the URL so we can safely know
   the version being downloaded.  However, if the source code is on Github,
   which it is in the case of `iperf3`, then `kraft` can figure this out for us.
   
   We can now use `kraft` to initialize a template library for us:

   ```bash
   cd ~/workspace/libs
   kraft lib init \
      --no-prompt \
      --author-name "Your Name" \
      --author-email "your@email.com" \
      --version 3.10.1 \
      --origin https://github.com/esnet/iperf \

>       iperf3

Author: Yes, you have access to an internal version of iperf3 so this will only apply to you and other contrbutors to Unikraft who can see this internal `lib-iperf3`: so it fails for me too.  The solve for you is to simply update your `.kraftrc` with only 1 `list.origins`, and that's the Unikraft core repo and then start this process of the tutorial.

```toml
[list]
origins = [
  "http://github.com/unikraft/unikraft.git",
]
```

   between Unikraft and the application in question (covered in the next section
   on making patches).

6. Now that we have added all the source files, let's try and build the
   application!  This step, again, usually occurs iteratively along with the
   previous step of adding a new file one-by-one.  Because the application has
   been `configure`d and we have `fetch`ed the contents, we can simply try
   running the build in the Unikraft application directory:

   ```bash
   cd ~/workspace/apps/iperf3
   kraft build
   ```
   
7. (Optional)  This step occurs less frequently, but is still useful to discuss
   in the context of porting an application to Unikraft.  Remember in [the
   Unikraft build lifecycle](#the-unikraft-build-lifecycle) that there is a step
   which occurs between fetching the remote origin code and compiling it. This
   step (3), known as `prepare`, is used to make modifications to the origin
   code before it is compiled.  This may be useful for applications which have
   complex build systems or auxiliary files which need to be created or modified
   before they are built.  Examples for `prepare`ing include:

>     * Running scripts which generate new source files from templates; 
>     * Compiling files preemptively before Unikraft starts `build`ing source
>       files;
>     * Checking for additional tools or building additional tools which are
>       required to build the library; and,
>     * Advanced patching techniques to the source files of the library which make
>       changes to it in a non-standard way.

Reviewer: ```suggestion
    * Running scripts which generate new source files from templates;
    * Compiling files preemptively before Unikraft starts `build`ing source
      files
    * Checking for additional tools or building additional tools which are
      required to build the library
    * Advanced patching techniques to the source files of the library which make
      changes to it in a non-standard way
```

   
   Preparation is done by adding Make targets to the `UK_PREPARE` variable:

   ```Makefile
   UK_PREPARE += mytarget
   ```

   Checking whether the library has been `prepare`d or adding a target which
   requires preparation before it can be executed is as simple as checking
   whether the following target exists:

   ```Makefile
   $(LIBIPERF3_BUILD)/.patched
   ```

   The `prepare` step is called naturally because of this target.  However, it
   can be called separately from `kraft` via:

   ```bash
   kraft prepare
   ```

The steps outlined above helped us begin the process of porting a simple
application to Unikraft.  It covers the major steps involved in the process of

> porting "from first principles," including addressing all the steps in the

Reviewer: ```suggestion
porting "from first principles", including addressing all the steps in the
```

construction lifecycle of Unikraft unikernels.

There are occasional caveats to this process, however.  This is to do with

> context of the "unikernel model," that is, single-purpose OSes with a single

Reviewer: ```suggestion
context of the "unikernel model", that is, single-purpose OSes with a single
```

address space, acting in a single process without context switches or costly
syscalls.  Applications developed for Linux userspace make a number of
assumptions about its runtime, for example:

>  * That all syscalls are available (which is not the case for Unikraft, although
>    there is significant work being done to bring more syscalls to Unikraft);
>  * That the filesystem is complete;
>  * That P in POSIX is _not_ silent: Unfortunately it is and *unix-type systems
>    do not always adhere to standards and make their own assumptions.  For
>    example, oftentimes there are differences between Linux and BSD-type OSes
>    which need to be accounted for; and,
>  * That all features are necessary.

Reviewer: ```suggestion
 * That all syscalls are available (which is not the case for Unikraft, although
   there is significant work being done to bring more syscalls to Unikraft)
 * That the filesystem is complete
 * That the P in POSIX is _not_ silent: Unfortunately it is and *unix-type systems
   do not always adhere to standards and make their own assumptions.  For
   example, oftentimes there are differences between Linux and BSD-type OSes
   which need to be accounted for
 * That all features are necessary
```

create one of these patches.

1. [The first
   patch](https://github.com/lancs-net/lib-iperf3/blob/staging/patches/0001-Fix-duplicate-import-of-netinet-tcp.h.patch)
   comes from an error which is thrown when compiling the `iperf_api.c` source
   file.  This file is 3rd to be compiled from the list of complete source
   files.  In this file, we are receiving a duplicate import of
   `<netinent/tcp.h>`, simply removing this import fixes it, so the patch
   addresses this issue.

2. [The second
   patch](https://github.com/lancs-net/lib-iperf3/blob/staging/patches/0002-Disable-SO_SNDBUF-and-SO_RCVBUF-checks.patch)
   comes as a result of [missing functionality from
   LwIP](https://github.com/lwip-tcpip/lwip/blob/b0e347158d8db640c6891f9f31f4e6d19dca200b/src/include/lwip/sockets.h#L220).
   The issue was discovered once the application was fully ported and was able
   to boot and run.  When the initialization sequence was on-going between the
   client and server of `iperf3`, it would crash during this sequence because
   LwIP does not support setting this option.  A patch was created simply to
   remove setting this option.  (Note: this may not be the most sensible

>    approach.)

Reviewer: ```suggestion
   approach)
```

   is once again to do with missing functionality from Unikraft.  In this case,
   the syscalls [`mmap`](https://linux.die.net/man/2/mmap) and
   [`munmap`](https://linux.die.net/man/2/munmap) are missing.  In this case,
   `iperf3`, used `mmap` simply to statically allocate a region of memory.  The
   trick used here is to simply replace instances of `mmap` with `malloc` and
   instances of `munmap` with `free`.

   **Note:** At the time writing this tutorial, [`mmap` and `munmap` are being
   actively worked on to be made available as syscalls in
   Unikraft](https://github.com/unikraft/unikraft/pull/247).

The above patches represent example use cases where patches may be necessary to
fix the application when bringing it to Unikraft.  The possibilities presented
in this tutorial are non-exhaustive, so take care.

The next section discusses in detail how to create a patch for the target
application or library.

### Preparing a patch for the application

When a change is identified and is to be provided as a patch to the application
or library during the compilation, it can be done using the procedure identified

> in this section.  Note that providing patches are an unfortunate workaround to
> the inherent differences between Linux userspace applicatios and libraries and
> unikernels.

Reviewer 2: ```suggestion
in this section.  Note that providing patches is an unfortunate workaround to
the inherent differences between Linux userspace applications and libraries and
unikernels.
```

Author: The "are" is correct, "patches" is plural. :)

   change you made and why.  This can be done through a number of successive
   steps, for example, as a result of having to make several changes to the
   application.

4. After your changes have been saved to the git log, export them as patches.
   For example, if you have made one (`1`) patch only, export it like so:

   ```bash
   git format-patch HEAD~1
   ```
  
   This will save a new `.patch` file in the current directory; which should be
   the origin source files of `iperf3`.

5. The next step is to create a `patches/` folder within the Unikraft port of
   the library and to move the new `.patch` file into this folder:

   ```bash
   mkdir ~/workspace/libs/iperf3/patches
   mv ~/workspace/apps/iperf3/build/libiperf3/origin/iperf-3.10.1/*.patch ~/workspace/libs/iperf3/patches
   ```

6. To register patches against Unikraft's build system such that they are
   applied before the compilation of all source files, simply indicate it in the
   library's `Makefile.uk`:

   ```Makefile
   # Add or edit ~/workspace/libs/iperf3/Makefile.uk
   LIBIPERF3_PATCHDIR = $(LIBIPERF3_BASE)/patches
   ```

This concludes the neccessary steps to port an application to Unikraft "from

> first princples"

Reviewer 2: ```suggestion
first principles"
```

## Workshop Exercises

### 01. Port `libfortune` to Unikraft

(Uni)kernel developers often seek guidance from elders, lost `man` pages,
wizards, source code comments and occassionally swear by the reproducible

> environment.  But the unfortunate truth is that "bitshifts happen," and we

Reviewer: ```suggestion
environment.  But the unfortunate truth is that "bitshifts happen", and we
```

cannot always [leverage guidance from mysterious
forces](https://www.ggbuddy.com/img/post/oymiswgv/midgvf/3245543291_ac5d471739d22f258621e4e7cbdcb98f_309e11564f795_m.jpg).

> A shared library called [`libfortune`](https://github.com/nderjung/libfortune)

Reviewer 2: Link https://github.com/nderjung/libfortune does not (yet?) seem to point to a valid repo. Maybe it's private?

Author: I need to make it public :)

Powered by
o.   .o       _ _               __ _
Oo   Oo  ___ (_) | __ __  __ _ ' _) :_
oO   oO ' _ `| | |/ /  _)' _` | |_|  _)
oOo oOO| | | | |   (| | | (_) |  _) :_
 OoOoO ._, ._:_:_,\_._,  .__,_:_, \___)
                   Tethys 0.5.0~825b115

"It always seems impossible until it is done."
        -- Nelson Mandela
```

### 02. Add fortunes to Unikraft's boot sequence

In this task, we are diving a little deeper into Unikraft's core and finding an
opportunity to meddle with internal features which can prove handy for certain
application contexts.  In this case, we are going to play with Unikraft's
extensible boot sequence to provide fortunes during the boot of an application.
After word got out, we found that everybody wanted fortunes, right before the
application started and `main()` was called.  This will provide the runtime of
the unikernel with good fortune and save it from crashes.

Unikraft calls various "constructor" (`ctor`) and "initialiser" (`init`)
methods during its boot sequence.  These constructors and initialisers are
located in a static section of the final binary image, `ctortab` and `inittab`,
respectively.  There are 7 entry points during the boot sequence:

> | Order | Level | Regitraring method                  | Type   |

Reviewer: ```suggestion
| Order | Level | Registering method                  | Type   |
```

constructors or initialisers at the same level.  This allows application
developers or library developers to correctly set up the unikernel by
registering a constructor or initialiser at the right time or before or after
others.

Initialisers have 6 different levels, allowing code to be injected before
certain operations occur during the boot sequence.  This includes, in order:
before and after the `plat`form drivers are initialised; before and after all
`lib`raries are initialised; before and after all filesystems (`rootfs`) are
initialised; and, before and after various "`sys`tem" methods are called.

The source code for this sequence is defined in [`ukboot`](https://github.com/unikraft/unikraft/blob/staging/lib/ukboot/boot.c).

In this task, add a new KConfig option to the Unikraft port of `libfortune`
which allows you to enable or disable the ability to introduce a fortune during
the bootsequence of a Unikernel.  Demonstrate the ability of using this library
by building the Unikraft port of `libfortune` to the Unikraft port of
[python3](https://github.com/unikraft/app-python3) and show a fortune before the
Unikraft banner.

### 03. Create a patch to introduce a new fortune

> There is a well-known kerel quote, which should be introduced to this library as

Reviewer: ```suggestion
There is a well-known kernel quote, which should be introduced to this library as
```

an easter egg for unikernel users:

"[Kernel hacking: where the time to solve a problem is inversely proportional to
the size of the resulting diff](https://twitter.com/avsm/status/1265666433914191876)"

  -- Anil Madhavapeddy

Please add this quote to `libfortune` as a patch so it is only available when

> used with unikernels.

Reviewer: The task to give feedback is missing!

   there is significant work being done to bring more syscalls to Unikraft);
 * That the filesystem is complete;
 * That P in POSIX is _not_ silent: Unfortunately it is and *unix-type systems
   do not always adhere to standards and make their own assumptions.  For
   example, oftentimes there are differences between Linux and BSD-type OSes
   which need to be accounted for; and,
 * That all features are necessary.

In the next section we address how we can make changes to the application before
it is compiled by the Unikraft build system in order to address the points
above.

## Invoking the application's `main` method

Traditionally, and by explicit design, Linux userspace code invokes a `main`
method (or symbol) for the start-of-execution of application logic.  Unikraft is
similar and invokes a [`weak`-ly attributed
symbol](https://gcc.gnu.org/onlinedocs/gcc-4.3.5/gcc/Function-Attributes.html)
for `main` in [its main
thread](https://github.com/unikraft/unikraft/blob/staging/lib/ukboot/boot.c#L75).
This is done so that it can be easily overwritten so as to invoke true
application-level functionality.  Without any `main` method, the unikernel will
simply boot and exit.

> All applications must follow the following standard prototype for `main`:

Reviewer: ```suggestion
All applications must implement the following standard prototype for `main`:
```

"follow the following" doesn't sound that good

during the pre-processing of the file which contains the method.  This flag
allows us to define macros in-line, and we can simply introduce a macro which
renames the `main` method to something else.

With `iperf3`, for example, we can rename the `main` method to `iperf3_main` by
adding a new library-specific `_FLAGS-y` entry in `Makefile.uk`:

```Makefile
LIBIPERF3_IPERF3_FLAGS-y += -Dmain=iperf3_main
```

The resulting object file for `main.c` will no longer include a symbol named
`main`.  At this point, when the final unikernel binary is linked, it will
simply quit.  We must now provide another `main` method.

To conditionally invoke the application's now renamed `main` method, it is
common to provide a new KConfig in the Unikraft library representing the port of
the application's `Config.uk` file, asking whether to "provide the main method".
For example, with `iperf3`:

```KConfig
if LIBIPERF3
config LIBIPERF3_MAIN_FUNCTION
	bool "Provide main function"
	default n
endif
```

When this option is enabled, we can either:

1. Disable the use of the `-D` flag as indicated above, conditionally in the
   `Makefile.uk`:
   ```Makefile

>    ifneq($(CONFIG_LIBIPERF3_MAIN_FUNCTION),y)

Reviewer: ```suggestion
   ifdef CONFIG_LIBIPERF3_MAIN_FUNCTION
```

I think it also works like this, whatever you prefer.

Author: I have not seen this syntax for make before

Reviewer: I used it a bit when porting p11-kit. Here is an example[1]:

[1] https://www.gnu.org/software/make/manual/html_node/Conditional-Syntax.html

Author: In this case, we want it to not equal `y`.  When providing main, in this case, we do not want to rewrite the method.
